vertx-grpc-server/src/main/java/io/vertx/grpc/server/impl/GrpcServerResponseImpl.java

```diff
@@ -234,12 +250,33 @@ private Future<Void> writeMessage(GrpcMessage message, boolean end) {
         responseTrailers.remove("grpc-message");
       }
       if (message != null) {
-        return httpResponse.end(GrpcMessageImpl.encode(message));
-      } else {
-        return httpResponse.end();
+        httpResponse.write(encodeMessage(message));
+      }
+      if (isGrpcWeb() && !trailersOnly) {
+        Buffer buffer = Buffer.buffer();
+        for (Map.Entry<String, String> trailer : responseTrailers) {
+          buffer.appendString(trailer.getKey())
+            .appendString(":")
+            .appendString(trailer.getValue())
+            .appendString("\r\n");
+        }
+        httpResponse.write(encodeMessage(new GrpcMessageImpl("identity", buffer, true)));
       }
+      return httpResponse.end();
     } else {
-      return httpResponse.write(GrpcMessageImpl.encode(message));
+      return httpResponse.write(encodeMessage(message));
+    }
+  }
+
+  private Buffer encodeMessage(GrpcMessage message) {
+    Buffer buffer = GrpcMessageImpl.encode(message);
+    if (contentType == GrpcMediaType.GRPC_WEB_TEXT_PROTO) {
+      return Buffer.buffer(Base64.getEncoder().encode(buffer.getBytes()));
```

Comment:
we should use io.netty.handler.codec.base64.Base64 here that encodes a ByteBuf to a ByteBuf which avoids ByteBuf -> byte[] -> byte[] -> ByteBuf

Reply:
Thanks for the tip